pilot/cmd/pilot-discovery/server/server.go

```diff
@@ -0,0 +1,494 @@
+// admit.ControllerOptions (other fields are filled out before constructing the admission controller). Only
+// used if running with k8s, Consul, or Eureka (not in a mock environment).
+type AdmissionArgs struct {
+	// ExternalAdmissionWebhookName is the name of the
+	// ExternalAdmissionHook which describes he external admission
+	// webhook and resources and operations it applies to.
+	ExternalAdmissionWebhookName string
+
+	// ServiceName is the service name of the webhook.
+	ServiceName string
+
+	// SecretName is the name of k8s secret that contains the webhook
+	// server key/cert and corresponding CA cert that signed them. The
+	// server key/cert are used to serve the webhook and the CA cert
+	// is provided to k8s apiserver during admission controller
+	// registration.
+	SecretName string
+
+	// Port where the webhook is served. Per k8s admission
+	// registration requirements this should be 443 unless there is
+	// only a single port for the service.
+	Port int
+
+	// RegistrationDelay controls how long admission registration
+	// occurs after the webhook is started. This is used to avoid
+	// potential races where registration completes and k8s apiserver
+	// invokes the webhook before the HTTP server is started.
+	RegistrationDelay time.Duration
+}
+
+// PilotArgs provides all of the configuration parameters for the Pilot discovery service.
+type PilotArgs struct {
+	DiscoveryOptions envoy.DiscoveryServiceOptions
+	Namespace        string
+	Mesh             MeshArgs
+	Config           ConfigArgs
+	Service          ServiceArgs
+	Admission        AdmissionArgs
+}
+
+// Server contains the runtime configuration for the Pilot discovery service.
+type Server struct {
+	config *pilotConfig
+}
+
+// NewServer creates a new Server instance based on the provided arguments.
+func NewServer(args PilotArgs, stop chan struct{}) (*Server, error) {
+	// If the namespace isn't set, try looking it up from the environment.
+	if args.Namespace == "" {
+		args.Namespace = os.Getenv("POD_NAMESPACE")
+	}
+
+	config := &pilotConfig{stop: stop}
+
+	// Apply the arguments to the configuration.
+	if err := initMesh(&args, config); err != nil {
+		return nil, err
+	}
+	if err := initKubeClient(&args, config); err != nil {
+		return nil, err
+	}
+	if err := initAdmissionController(&args, config); err != nil {
+		return nil, err
+	}
+	if err := initMixerSan(&args, config); err != nil {
+		return nil, err
+	}
+	if err := initConfigController(&args, config); err != nil {
+		return nil, err
+	}
+	if err := initServiceControllers(&args, config); err != nil {
+		return nil, err
+	}
+	if err := initDiscoveryService(&args, config); err != nil {
+		return nil, err
+	}
+	return &Server{config: config}, nil
+}
+
+// Run starts the pilot discovery service.
+func (s Server) Run() {
+	// Now start all of the services.
+	for _, runFn := range s.config.runFuncs {
+		runFn()
```

Comment:
Looks like every run function is a goroutine. Change to `go runFn()` and make each function blocking?

Reply:
A couple of the functions are non-blocking (e.g. the monitor)